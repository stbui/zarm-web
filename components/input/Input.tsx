import React, { Component } from 'react';
import { InputProps, InputCoreProps, TextAreaProps } from './PropsType';
import Textarea from './Textarea';
import InputCore from './InputCore';
import Group from './Group';

class Input extends Component<InputProps, any> {
  static defaultProps = {
    prefixCls: 'zw-input',
    type: 'text',
    shape: 'radius',
    bordered: 'underline',
  };

  static Group: typeof Group;

  private input;

  focus() {
    if (this.input) {
      this.input.focus();
    }
  }

  blur() {
    if (this.input) {
      this.input.blur();
    }
  }

  render() {
    const { type } = this.props;
    let children: React.ReactNode;
    if (type === 'textarea') {
      children = <Textarea ref={(el) => { this.input = el; }} {...this.props as TextAreaProps} />;
    } else {
      children = <InputCore ref={(el) => { this.input = el; }} {...this.props as InputCoreProps} />;
    }

    return children;
  }
}

export default Input;
