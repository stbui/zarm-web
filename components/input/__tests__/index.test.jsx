import React from 'react';
import { render, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Input from '../index';

describe('Input', () => {
  it('renders different types of Input correctly', () => {
    const wrapper = render(
      <div>
        <Input type="text" placeholder="请输入" />
        <Input type="textarea" placeholder="请输入" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders different sizes of Input correctly', () => {
    const wrapper = render(
      <div>
        <Input size="lg" placeholder="请输入" />
        <Input size="md" placeholder="请输入" />
        <Input size="sm" placeholder="请输入" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Input with radius correctly', () => {
    const wrapper = render(
      <div>
        <Input shape="radius" type="text" placeholder="请输入" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Disabled Input correctly', () => {
    const wrapper = render(
      <div>
        <Input disabled type="text" value="禁用" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with initial value', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <div>
        <Input onChange={onChange} value="hey" type="text" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when changing Input value', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Input onChange={onChange} type="text" />
      </div>,
    );

    wrapper.find('input').simulate('change');
    expect(onChange).toHaveBeenCalled();
  });

  it('should trigger clear event correctly when click clear icon', () => {
    let argumentEventObject;
    let argumentEventObjectValue;
    const onChange = (e) => {
      argumentEventObject = e;
      argumentEventObjectValue = e.target.value;
    };
    const wrapper = mount(<Input clearable value="111" onChange={onChange} />);
    wrapper
      .find('.zw-input__clear-icon')
      .at(0)
      .simulate('click');
    expect(argumentEventObject.type).toBe('click');
    expect(argumentEventObjectValue).toBe('');
    expect(
      wrapper
        .find('input')
        .at(0)
        .getDOMNode().value,
    ).toBe('111');
  });

  it('should focus input after clear', () => {
    const wrapper = mount(<Input clearable defaultValue="111" />);
    wrapper
      .find('.zw-input__clear-icon')
      .at(0)
      .simulate('click');
    expect(document.activeElement).toBe(
      wrapper
        .find('input')
        .at(0)
        .getDOMNode(),
    );
  });

  it('should not support clearable when it is disabled', () => {
    const wrapper = mount(<Input clearable defaultValue="111" disabled />);
    expect(wrapper.find('.zw-input__clear-icon').length).toBe(0);
  });
});
