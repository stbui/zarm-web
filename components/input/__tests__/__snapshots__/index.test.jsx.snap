// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders Disabled Input correctly 1`] = `
<div>
  <div
    class="zw-input zw-input--md zw-input--radius zw-input--disabled zw-input--underline"
  >
    <input
      class="zw-input__inner"
      disabled=""
      type="text"
      value="禁用"
    />
    <div>
      <div
        class="zw-input__line"
      />
      <div
        class="zw-input__focus-line"
      />
    </div>
  </div>
</div>
`;

exports[`Input renders Input with radius correctly 1`] = `
<div>
  <div
    class="zw-input zw-input--md zw-input--radius zw-input--underline"
  >
    <input
      class="zw-input__inner"
      placeholder="请输入"
      type="text"
      value=""
    />
    <div>
      <div
        class="zw-input__line"
      />
      <div
        class="zw-input__focus-line"
      />
    </div>
  </div>
</div>
`;

exports[`Input renders correctly with initial value 1`] = `
<div>
  <Input
    bordered="underline"
    onChange={[MockFunction]}
    prefixCls="zw-input"
    shape="radius"
    type="text"
    value="hey"
  />
</div>
`;

exports[`Input renders different sizes of Input correctly 1`] = `
<div>
  <div
    class="zw-input zw-input--lg zw-input--radius zw-input--underline"
  >
    <input
      class="zw-input__inner"
      placeholder="请输入"
      type="text"
      value=""
    />
    <div>
      <div
        class="zw-input__line"
      />
      <div
        class="zw-input__focus-line"
      />
    </div>
  </div>
  <div
    class="zw-input zw-input--md zw-input--radius zw-input--underline"
  >
    <input
      class="zw-input__inner"
      placeholder="请输入"
      type="text"
      value=""
    />
    <div>
      <div
        class="zw-input__line"
      />
      <div
        class="zw-input__focus-line"
      />
    </div>
  </div>
  <div
    class="zw-input zw-input--sm zw-input--radius zw-input--underline"
  >
    <input
      class="zw-input__inner"
      placeholder="请输入"
      type="text"
      value=""
    />
    <div>
      <div
        class="zw-input__line"
      />
      <div
        class="zw-input__focus-line"
      />
    </div>
  </div>
</div>
`;

exports[`Input renders different types of Input correctly 1`] = `
<div>
  <div
    class="zw-input zw-input--md zw-input--radius zw-input--underline"
  >
    <input
      class="zw-input__inner"
      placeholder="请输入"
      type="text"
      value=""
    />
    <div>
      <div
        class="zw-input__line"
      />
      <div
        class="zw-input__focus-line"
      />
    </div>
  </div>
  <div
    class="zw-input zw-input--textarea zw-input--radius zw-input--bordered"
  >
    <textarea
      class="zw-input__inner"
      placeholder="请输入"
      type="textarea"
    />
  </div>
</div>
`;
